Ukol_2/cpp/sifry.cpp

```diff
@@ -1,47 +1,228 @@
+        if (octalStr[i] == '\\') {
+
+            string octalValue = octalStr.substr(i + 1, 3);
+            int asciiValue = stoi(octalValue, nullptr, 8);
+
+
+            result += static_cast<char>(asciiValue);
+
+            i += 4;
+        } else {
+
+            result += octalStr[i];
+            i++;
+        }
+    }
+
+    return result;
 }
 
+
+
 // Funkce pro Caesarovu šifru
 std::string caesar_sifra(const std::string &text, int posun, bool sifrovat)
 {
+
   // Implementace Caesarovy šifry
   // sifrovat = true pro šifrování, sifrovat = false pro dešifrování
-  return "";
+  string lowerCaseText = text;
+  transform(lowerCaseText.begin(),lowerCaseText.end(), lowerCaseText.begin(), ::tolower);
+  
+
+  if(posun<0){
+    posun = 26 + (posun%26); 
+  }else if(posun>26){
+    posun = posun % 26;
+  }
+
+
+  if(sifrovat){
+    char* charArray = new char[text.length()+1];
+    for(int i = 0; i < text.length(); i++){
+      if(lowerCaseText[i]+posun<=122){
+        charArray[i] = lowerCaseText[i] + posun;
+      }else{
+        charArray[i] = lowerCaseText[i] + posun - 26;
+      }
+    }
+    string str(charArray);
+    return charArray;
+  }else{
+    char* charArray = new char[lowerCaseText.length()+1];
+    for(int i = 0; i < lowerCaseText.length(); i++){
+      charArray[i] = lowerCaseText[i] - posun;
+    }
+    string str(charArray);
+    return charArray;
+  }  
 }
 
+
+
 // Funkce pro Vigenerovu šifru
 std::string vigener_sifra(const std::string &text, const std::string &klic, bool sifrovat)
 {
-  // Implementace Vigenerovy šifry
-  // sifrovat = true pro šifrování, sifrovat = false pro dešifrování
-  return "";
+
+  string lowerCaseText = text;
+  transform(lowerCaseText.begin(),lowerCaseText.end(), lowerCaseText.begin(), ::tolower);
+  string lowerCaseKey = klic;
+  transform(lowerCaseKey.begin(),lowerCaseKey.end(), lowerCaseKey.begin(), ::tolower);
+
+  char* charArray = new char[lowerCaseText.length()+1];
+  int j = 0;
+  int keyLength = klic.length();
+
+  if(sifrovat){
+      for(int i = 0; i < lowerCaseText.length();i++){
+        if(lowerCaseText[i] + lowerCaseKey[i] - 97 < 122){
+          charArray[i] = lowerCaseText[i] + lowerCaseKey[j] - 97;
+        }else{
+          charArray[i] = lowerCaseText[i] + lowerCaseKey[j] - 123;
+        }
+        j++;
+        if(j>keyLength){
+          j = 0;
+        }
+      }  
+      string str(charArray);
+      return charArray;
+  }else{
+      for(int i = 0; i < lowerCaseText.length();i++){
+        if(lowerCaseText[i] - lowerCaseKey[i] + 97 > 96){
+          charArray[i] = lowerCaseText[i] - lowerCaseKey[j] + 97;
+        }else{
+          charArray[i] = lowerCaseText[i] - lowerCaseKey[j] + 123;
+        }
+        j++;
+        if(j>keyLength){
+          j = 0;
+        }
+      }
+      string str(charArray);
+      return charArray;
+  }
 }
 
 // Funkce pro XOR šifru
 std::string xor_sifra(const std::string &text, const std::string &klic, bool sifrovat)
```

Comment:
XOR mi taky nevraci uplně ty spravne hodnoty ale tam mě to nepřekvapuje hodně zalezi na implementaci teto funkce.